### ISSUE TITLE
Threaded loading crashes/general issues in 3.2 (and possibly 4.0)

### ISSUE THREAD

**Issue by EIREXE** [godotengine/godot#44606]
**Godot version:** 68013d23932688e57b489600f4517dd280edc464

**OS/device including version:**
Windows, Linux, GLES3, nvidia and AMD GPUs

**Issue description:**
Threaded loading in general seems unstable, particularly when loading images and scenes.

The problems seem to be:

- Memory corruption as reported by @Xananax (although I haven't been able to reproduce it), he stated that separation of memory between threads doesn't seem to work properly even when mutexes are used.
- Crashes/hangs.
- Textures not loading properly.

On an important note, I vividly remember doing threading in Godot 3.0 without problems, so I suspect the issue was introduced after that.

**Steps to reproduce:**
Develop a program that uses multi threading to load the aforementioned resources in threads.

**Minimal reproduction project:**
I've been unable to reproduce crashes (yet) but linked is a project that shows some issues on my machine, mainly it appears like it's only able to load some of the scenes, scenes 1 2 3 and 4 are identical.

![imagen](https://user-images.githubusercontent.com/3101690/102969532-264a9e00-44f6-11eb-8e88-04bf23c05972.png)

4 is loaded properly I assume because it's the last one.

2 also loads properly because the load has been increased by including a few audio stream players with an audio stream, so maybe there's some timing issue there.

Using multi threaded rendering mode seems to fix this issue in particular, but in my experience the crashes remain.

I suspect this might be related to allocating textures outside the main thread, as in my experience loading images as Image resources and only putting them in ImageTextures once they are in the main thread seems to solve the issue (additionally, I did some printf debugging that seems to confirm it's possible for textures to reach the main thread without being allocated, even when using VisualServer.force_sync).

This is all I've got for now, I'm still attempting to reproduce the crashes in a project, but it may be that they only happen with an increased load, so I've laid out the most probable explanations I can think of.

This is by the way entirely unrelated to the new 4.0 APIs that allow threaded loading of resources to be managed by the engine which I suspect suffer from the same problems, but I haven't tested this in 4.0.

[ThreadedLoadingExample.zip](https://github.com/godotengine/godot/files/5733782/ThreadedLoadingExample.zip)

**Comment by Xananax** [comment]
In a semi large project we're doing, we've been receiving random object when loading in threads. We would be loading, for example two scenes in a thread, and another scene in another, and we'd receive some random node or object, which isn't at all what is expected.      
Since the project is rather big and complex, I cannot conclusively rule out that we may be doing something wrong, but it certainly doesn't look that way, which is why I'll be following this issue. 

This is not consistent nor easy to reproduce, which is why no bug was reported from out side.

It also seems to be dependent on the architecture, because we have one beta tester to whom it happens semi-consistently, whereas it only ever happened once for me.  

Because of the randomness and lack of reproducibility, it seems that a racing condition is happening. Because a valid object is returned (i.e, a node) but not the one asked for, I suspect that data is somehow being shared between threads.

**Comment by EIREXE** [comment]
> It also seems to be dependent on the architecture, because we have one beta tester to whom it happens semi-consistently, whereas it only ever happened once for me.

I report similar results, I have not had the game crash on me a lot, while for some users it happens all the time. we use different OSes and CPU architectures too.

**Comment by trommlbomml** [comment]
This really sounds like some unproper memory handling. If hard to reproduce you can only look at the code and think where concurrent access breaks it. I remember OpenGL drivers also may not be stable to be called from different threads, even with proper contexts (but some years ago...)

**Comment by RandomShaper** [comment]
Some time ago I made this PR to try to make threaded loading safe: https://github.com/godotengine/godot/pull/40765

Unfortunately, the project I developed that for still had crashes related to threaded loading, only much harder to reproduce. Therefore, I guess that PR helped, but in the end we had to switch to "polled-yielded" loading. Anyway, those building Godot themselves, could give that PR a try, but, I insist, be aware that it may give the impression of fixing the problem at first and later you get crashes again.

Depending on the reason for using threaded loading in your project, you may want to did the same we did or try to use that PR and report crashes so we can understand better what's missing in Godot for thread safety.

**Comment by Calinou** [comment]
> I remember OpenGL drivers also may not be stable to be called from different threads, even with proper contexts (but some years ago...)

Indeed, OpenGL isn't really thread-safe to begin with (unlike Vulkan).

**Comment by RandomShaper** [comment]
If a project is dealing with the `VisualServer` from multiple threads, it must do some locking logic or use single-safe or multithreaded renderer threading models.

However, that's not enough, because there are some other sources or synchronization problems in nodes and resources.

**Comment by EIREXE** [comment]
> > I remember OpenGL drivers also may not be stable to be called from different threads, even with proper contexts (but some years ago...)
> 
> Indeed, OpenGL isn't really thread-safe to begin with (unlike Vulkan).

However I vividly remember loading scenes and instancing them in threads not causing crashes in the past (around 3.0) I am not at home so I can't find what version of the engine causes the issues but I'll try.

**Comment by ICatRegister** [comment]
Couldn't reproduce crashes or freezes in the project.
Also after deleting the line VisualServer.force_sync(), all 4 scenes are loaded without issues

**Comment by EIREXE** [comment]
> Couldn't reproduce crashes or freezes in the project.
> Also after deleting the line VisualServer.force_sync(), all 4 scenes are loaded without issues

Yes the example project wasn't supposed to depict that, I haven't been able to make it happen in isolation (yet)

Also interesting the visualserver force sync is causing part of the problem, it doesn't make much sense hmm

**Comment by EIREXE** [comment]
After some investigation I've discovered this might be related to interacting with the visualserver.

The way I reached this conclusion is that I deployed a build of the game using ANGLE + Vulkan/d3d11, and suddenly the crashes were gone for some users that had them happen all the time.

This leads me to believe the crash originates in OpenGL drivers somehow, since OpenGL is not thread safe and maybe ANGLE is, does it make sense to everyone else?

**Comment by RandomShaper** [comment]
That makes sense if the rendering thread model is not multi-threaded (project setting).

An OpenGL context is only guaranteed to work well from the thread it was made current for. (I guess some drivers may have their own sync mechanism, so they are safe even if using the same context from different threads, but that's not something to rely on. For what it's worth, ANGLE only does locking if you set multiple contexts for resource sharing.)

**Comment by RandomShaper** [comment]
I just realized I said mostly the same 10 days ago. I didn't remember I already participated in this discussion. 😅

**Comment by EIREXE** [comment]
> 
> 
> I just realized I said mostly the same 10 days ago. I didn't remember I already participated in this discussion. 😅

Maybe the documentation should be updated? or shouldn't the force_sync method be enough?

**Comment by trommlbomml** [comment]
Wouldnt it be good to have a warning when accessing the visual server from another thread in these cases?

**Comment by EIREXE** [comment]
I have to say that after some investigation it appears I jumped the gun, using ANGLE makes it happen less but it does still happen, so I believe the issue is not just fixed with force_sync()

**Comment by tcoxon** [comment]
We've run into this too. Godot 3.5.1, GLES3, Windows & Linux, AMD & NVIDIA. We (have tried to) load scenes on a background thread, so that loading screens can update smoothly.

In our case we get graphical glitches and crashes that are impossible to reproduce, and sometimes an error like this is logged during loading (in release builds only):

```
ERROR: Index p_surface = 0 is out of bounds (instance->materials.size() = 0).
   at: instance_set_surface_material (servers/visual/visual_server_scene.cpp:1131)
```

If I set the graphics thread model to Multi-Threaded that error goes away (but according to the original description this won't fix the fundamental problem, and reportedly crashes remain).

If there is not currently a plan to fix this, can I suggest documenting that resources cannot be safely loaded on a thread? The docs still say that [ResourceInteractiveLoader can be used from multiple threads](https://docs.godotengine.org/en/stable/tutorials/io/background_loading.html#using-multiple-threads).
